trie.h

```diff
@@ -0,0 +1,37 @@
+#ifndef TRIE_H
+#define TRIE_H
+
+#include <map>
+#include <vector>
+#include <string>
+#include <memory>
+
+using std::string;
+using std::map;
+using std::vector;
+using std::shared_ptr;
+
+class Trie {
+private:
+    struct Node {
+        map<char, shared_ptr<Node>> children;
+        bool value;
+    };
+
+    shared_ptr<Node> head;
+
```

Comment:
Yeah, I thought this was gross too. I recalled pimpl at the time and looked it up wondering if I should do it. So which of those options would be the best for this case? Maybe I'll look at what STL does.